crates/std_detect/src/detect/cache.rs

```diff
@@ -124,61 +133,82 @@ impl Cache {
 #[cfg(target_pointer_width = "32")]
 struct Cache(AtomicU32, AtomicU32);
 
+#[cfg(target_pointer_width = "32")]
+const UNINITIALIZED_VALUE: u32 = u32::max_value();
+
+#[cfg(target_pointer_width = "32")]
+const UNINITIALIZED_BIT: u32 = 31;
+
+/// Maximum number of features that can be cached.
+/// We reserve a single bit in each of the two atomic values
+/// to check if the cache is initialized
+#[cfg(target_pointer_width = "32")]
+const CACHE_CAPACITY: u32 = 62;
+
 #[cfg(target_pointer_width = "32")]
 impl Cache {
     /// Creates an uninitialized cache.
     const fn uninitialized() -> Self {
         Cache(
-            AtomicU32::new(u32::max_value()),
-            AtomicU32::new(u32::max_value()),
+            AtomicU32::new(UNINITIALIZED_VALUE),
+            AtomicU32::new(UNINITIALIZED_VALUE),
         )
     }
-    /// Is the cache uninitialized?
-    #[inline]
-    pub(crate) fn is_uninitialized(&self) -> bool {
-        self.1.load(Ordering::Relaxed) == u32::max_value()
-    }
 
     /// Is the `bit` in the cache set?
     #[inline]
-    pub(crate) fn test(&self, bit: u32) -> bool {
-        if bit < 32 {
-            test_bit(CACHE.0.load(Ordering::Relaxed) as u64, bit)
+    fn test(&self, bit: u32) -> bool {
+        if bit < 31 {
+            let cache_value = CACHE.0.load(Ordering::Relaxed);
+            if !test_bit(cache_value, UNINITIALIZED_BIT) {
+                test_bit(cache_value, bit)
+            } else {
+                initializer(bit)
+            }
         } else {
-            test_bit(CACHE.1.load(Ordering::Relaxed) as u64, bit - 32)
+            let cache_value = CACHE.1.load(Ordering::Relaxed);
+            if !test_bit(cache_value, UNINITIALIZED_BIT) {
+                test_bit(cache_value, bit - 31)
+            } else {
+                initializer(bit)
+            }
         }
     }
 
     /// Initializes the cache.
     #[inline]
-    pub(crate) fn initialize(&self, value: Initializer) {
-        let lo: u32 = value.0 as u32;
-        let hi: u32 = (value.0 >> 32) as u32;
+    fn initialize(&self, value: Initializer) {
+        let lo: u32 = unset_bit(value.0, UNINITIALIZED_BIT) as u32;
+        let hi: u32 = unset_bit(value.0 >> 31, UNINITIALIZED_BIT) as u32;
         self.0.store(lo, Ordering::Relaxed);
         self.1.store(hi, Ordering::Relaxed);
```

Comment:
Would it be possible here to have the `self.0.store` happen before the `self.1.store` instead of doing all other changes ?

Reply:
No, I don' think that would be enough. As far as I can tell, you need to use `Release` and `Acquire` semantics to solve it this way.